tst/ExponentialBackoffUtilsTest.cpp

```diff
@@ -0,0 +1,232 @@
+    EXPECT_EQ(5, pExponentialBackoffState->pExponentialBackoffConfig->maxRetryCount);
+    EXPECT_EQ(10, pExponentialBackoffState->pExponentialBackoffConfig->maxWaitTime);
+    EXPECT_EQ(20, pExponentialBackoffState->pExponentialBackoffConfig->retryFactorTime);
+    EXPECT_EQ(1, pExponentialBackoffState->pExponentialBackoffConfig->jitterFactor);
+
+    freeExponentialBackoffState(&pExponentialBackoffState);
+    EXPECT_TRUE(pExponentialBackoffState == NULL);
+    SAFE_MEMFREE(pExponentialBackoffConfig);
+}
+
+/**
+ * This test validates un-bounded exponentially backed-off retries with an upper
+ * bound on the actual retry wait time.
+ * The test performs 7 consecutive and validates the wait time for each retry.
+ * The retry config has an upper limit on
+ * The test verifies reception of error upon calling exponentialBackoffBlockingWait after
+ * configured retries are exhausted.
+ *
+ * This test takes roughly 15 seconds to execute
+ */
+TEST_F(ExponentialBackoffUtilsTest, testExponentialBackoffBlockingWait_Unbounded)
+{
+    PExponentialBackoffState pExponentialBackoffState = NULL;
+    PExponentialBackoffConfig pExponentialBackoffConfig = NULL;
+    pExponentialBackoffConfig = (PExponentialBackoffConfig) MEMALLOC(SIZEOF(ExponentialBackoffConfig));
+    EXPECT_TRUE(pExponentialBackoffConfig != NULL);
+
+    getTestRetryConfiguration(pExponentialBackoffConfig);
+
+    EXPECT_EQ(STATUS_SUCCESS, initializeExponentialBackoffState(&pExponentialBackoffState, pExponentialBackoffConfig));
+    EXPECT_TRUE(pExponentialBackoffState != NULL);
+    EXPECT_TRUE(pExponentialBackoffState->pExponentialBackoffConfig != NULL);
+
+    std::vector<double> actualRetryWaitTime;
+    std::vector<Range> acceptableRetryWaitTimeRangeMilliSec {
+        {200.0, 300.0},     // 1st retry 200ms + jitter
+        {400.0, 500.0},     // 2st retry 400ms + jitter
+        {800.0, 900.0},     // 3st retry 800ms + jitter
+        {1600.0, 1700.0},   // 4st retry 1600ms + jitter
+        {3000.0, 3100.0},   // 5st retry 3000ms + jitter
+        {3000.0, 3100.0},   // 6st retry 3000ms + jitter
+        {3000.0, 3100.0}    // 7st retry 3000ms + jitter
+    };
+
+    for (int retryCount = 0; retryCount < 7; retryCount++) {
+        double currentTimeMilliSec = GETTIME()/(HUNDREDS_OF_NANOS_IN_A_MILLISECOND * 1.0);
+        EXPECT_EQ(STATUS_SUCCESS, exponentialBackoffBlockingWait(pExponentialBackoffState));
+        validateExponentialBackoffWaitTime(
+                pExponentialBackoffState,
+                currentTimeMilliSec,
+                retryCount+1,
+                BACKOFF_IN_PROGRESS,
+                acceptableRetryWaitTimeRangeMilliSec.at(retryCount));
+    }
+
+    // We have configured minTimeToResetRetryState as 5 seconds. So sleep for just over 5 seconds
+    // and then verify the next retry has retry count as 1
+    std::this_thread::sleep_for(std::chrono::milliseconds(5100));
```

Comment:
You can avoid sleeping in the tests by overriding the GETTIME functionality. Most of the unit tests drive at the CPU speed with synthetic timers. Same can be said about the randomness for the jitter tests.

Reply:
I'll check that